File: utils/mock_plugin.py
#!/usr/bin/env python3
# ===-----------------------------------------------------------------------===#
#
# This source file is part of the Swift.org open source project
#
# Copyright (c) 2023 Apple Inc. and the Swift project authors
# Licensed under Apache License v2.0 with Runtime Library Exception
#
# See https://swift.org/LICENSE.txt for license information
# See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
#
# ===-----------------------------------------------------------------------===#
#
# Helper python module to write a mock compiler plugin for testing compiler
# plugin mechanism. Test plugins can be like
#
#   import sys
#   sys.path.append('/path/to/utils')
#   import mock_plugin
#
#   mock_plugin.TEST_SPEC = [ ... ]
#   mock_plugin.main()
#
# where 'TEST_SPEC' is a list of pairs of an expected request and the result to
# respond. For example:
#   {'expect': {'testRequest': {'params': []}},
#    'response': {'testRequestResult': [1,2,"foo"]}}
# this spec matches a request that is an object with 'testRequest' key with an
# object with 'params' key with an empty array. When the host system sends a
# request matching, this mock plugin replies to it with the 'response' object.
# ===-----------------------------------------------------------------------===#

import json
import struct
import sys

# This should be filled by the client.
TEST_SPEC = []


def match(value, expect):
    if isinstance(value, dict) and isinstance(expect, dict):
        # For dictionary, only check the expected keys. i.e. the value can
        # contain extra items.
        for key, expectVal in expect.items():
            if key not in value:
                return False
            if not match(value[key], expectVal):
                return False
        return True

    if isinstance(value, list) and isinstance(expect, list):
        # Check all elements.
        if len(value) != len(expect):
            return False
        for element, expectVal in zip(value, expect):
            if not match(element, expectVal):
                return False
        return True

    # Otherwise, just check the equality.
    return value == expect


def substitute(value, req):
    if isinstance(value, str):
        result = value.format(req=req)
        # for "={req[...]}" format, covert it to integer.
        if value.startswith('={') and value.endswith("}"):
            return int(result[1:])
        return result

    if isinstance(value, dict):
        result = {}
        for key, value in value.items():
            result[key] = substitute(value, req)
        return result

    if isinstance(value, list):
        result = []
        for value in value:
            result.append(substitute(value, req))
        return result

    return value


def handle_request(req):
    for spec in TEST_SPEC:
        if not match(req, spec['expect']):
            continue
        if spec.get('handled', False):
            raise ValueError('already handled request')
        spec['handled'] = True
        return substitute(spec['response'], req)

    raise ValueError('could not find matching request')


def main():
    # Message handling loop.
    while True:
        # Read request
        request_header = sys.stdin.buffer.read(8)
        if len(request_header) < 8:
            break
        request_size = struct.unpack('<Q', request_header)[0]
        request_data = sys.stdin.buffer.read(request_size)
        if len(request_data) != request_size:
            break
        request_object = json.loads(request_data)

        # Handle request
        response_object = handle_request(request_object)

        # Write response
        response_json = json.dumps(response_object)
        response_data = response_json.encode('utf-8')
        response_size = len(response_data)
        response_header = struct.pack('<Q', response_size)
        sys.stdout.buffer.write(response_header)
        sys.stdout.buffer.write(response_data)
        sys.stdout.buffer.flush()


if __name__ == '__main__':
    main()

File: tools/libMockPlugin/MockPlugin.cpp
//===--- MockPlugin.cpp ---------------------------------------------------===//
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2023 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
//
//===----------------------------------------------------------------------===//

#include "swift-c/MockPlugin/MockPlugin.h"
#include "llvm/ADT/StringExtras.h"
#include "llvm/Support/Endian.h"
#include "llvm/Support/JSON.h"

#include <stdio.h>

namespace {
struct TestItem {
  llvm::json::Value expect;
  llvm::json::Value response;
  bool handled = false;

  TestItem(const llvm::json::Value expect, const llvm::json::Value response)
      : expect(expect), response(response) {}
};

struct TestRunner {
  std::vector<TestItem> items;

  TestRunner() {}
  static std::unique_ptr<TestRunner> create(const char *testSpecStr);

  llvm::json::Value createResponse(const TestItem &item,
                                   const llvm::json::Value &req);

  TestItem *findMatchItem(const llvm::json::Value &req);

  int run();
};
} // namespace

std::unique_ptr<TestRunner> TestRunner::create(const char *testSpecStr) {
  // Read test spec.
  auto testSpec = llvm::json::parse(testSpecStr);
  if (!testSpec) {
    llvm::errs() << "failed to parse test spec JSON\n";
    llvm::handleAllErrors(testSpec.takeError(),
                          [](const llvm::ErrorInfoBase &err) {
                            llvm::errs() << err.message() << "\n";
                          });
    return nullptr;
  }

  auto *items = testSpec->getAsArray();
  if (!items) {
    llvm::errs() << "test spec list must be an array\n";
    return nullptr;
  }

  std::unique_ptr<TestRunner> runner(new TestRunner());

  for (auto item : *items) {
    auto *obj = item.getAsObject();
    if (!obj) {
      llvm::errs() << "test spec must be an object\n";
      return nullptr;
    }
    auto *expect = obj->get("expect");
    auto *response = obj->get("response");
    if (!expect || !response) {
      llvm::errs() << "test item must have 'expect' and 'response'\n";
      return nullptr;
    }
    runner->items.emplace_back(*expect, *response);
  }

  return runner;
}

/// Replace string values starting with '=req' (e.g. '"=req.foo.bar[2].baz"') in
/// \p value with the corresponding values from \p req .
static const llvm::json::Value substitute(const llvm::json::Value &value,
                                          const llvm::json::Value &req) {

  // Recursively substitute objects and arrays.
  if (auto *obj = value.getAsObject()) {
    llvm::json::Object copy;
    for (auto &kv : *obj) {
      copy[kv.first] = substitute(kv.second, req);
    }
    return copy;
  }
  if (auto *arr = value.getAsArray()) {
    llvm::json::Array copy;
    for (auto &elem : *arr) {
      copy.push_back(substitute(elem, req));
    }
    return copy;
  }

  auto str = value.getAsString();
  if (!str || !str->starts_with("=req")) {
    // Not a substitution.
    return value;
  }

  auto path = str->substr(4);
  const llvm::json::Value *subst = &req;
  while (!path.empty()) {
    // '.' <alphanum> -> object key.
    if (path.starts_with(".")) {
      if (subst->kind() != llvm::json::Value::Object)
        return "<substitution error: not an object>";
      path = path.substr(1);
      auto keyLength =
          path.find_if([](char c) { return !llvm::isAlnum(c) && c != '_'; });
      auto key = path.slice(0, keyLength);
      subst = subst->getAsObject()->get(key);
      path = path.substr(keyLength);
      continue;
    }
    // '[' <digit>+ ']' -> array index.
    if (path.starts_with("[")) {
      if (subst->kind() != llvm::json::Value::Array)
        return "<substitution error: not an array>";
      path = path.substr(1);
      auto idxlength = path.find_if([](char c) { return !llvm::isDigit(c); });
      size_t idx;
      (void)path.slice(0, idxlength).getAsInteger(10, idx);
      subst = &(*subst->getAsArray())[idx];
      path = path.substr(idxlength);
      if (!path.starts_with("]"))
        return "<substitution error: missing ']' after digits>";
      path = path.substr(1);
      continue;
    }
    // Malformed.
    return "<substitution error: malformed path>";
  }
  return *subst;
}

llvm::json::Value TestRunner::createResponse(const TestItem &item,
                                             const llvm::json::Value &req) {
  auto response = item.response;
  return substitute(response, req);
}

/// Check if \p expect matches \p val .
static bool match(const llvm::json::Value &expect,
                  const llvm::json::Value &val) {
  if (expect.kind() != val.kind())
    return false;

  switch (expect.kind()) {
  case llvm::json::Value::Object: {
    auto *exp = expect.getAsObject();
    auto *obj = val.getAsObject();
    return llvm::all_of(*exp, [obj](const auto &kv) {
      auto *val = obj->get(kv.first.str());
      return val && match(kv.second, *val);
    });
    return true;
  }

  case llvm::json::Value::Array: {
    auto *exp = expect.getAsArray();
    auto *arr = val.getAsArray();
    if (exp->size() != arr->size())
      return false;
    return llvm::all_of_zip(
        *exp, *arr,
        [](const auto &expect, const auto &val) { return match(expect, val); });
  }

  default:
    return expect == val;
  }
}

TestItem *TestRunner::findMatchItem(const llvm::json::Value &req) {
  for (auto &item : items) {
    if (match(item.expect, req)) {
      return &item;
    }
  }
  return nullptr;
}

int TestRunner::run() {
  size_t ioSize;
  while (true) {
    // Read request header.
    uint64_t request_header;
    ioSize = fread(&request_header, sizeof(request_header), 1, stdin);
    if (!ioSize) {
      // STDIN is closed.
      return 0;
    }

    // Read request data.
    auto request_size = llvm::support::endian::byte_swap(
        request_header, llvm::endianness::little);
    llvm::SmallVector<char, 0> request_data;
    request_data.assign(request_size, 0);
    ioSize = fread(request_data.data(), request_size, 1, stdin);
    if (!ioSize) {
      llvm::errs() << "failed to read request data\n";
      return 1;
    }

    // Parse request object.
    auto request_obj =
        llvm::json::parse({request_data.data(), request_data.size()});
    if (!request_obj) {
      llvm::handleAllErrors(
          request_obj.takeError(),
          [](llvm::ErrorInfoBase &err) { llvm::errs() << err.message(); });
      return 1;
    }

    // Handle
    auto *item = findMatchItem(*request_obj);
    if (!item) {
      llvm::errs() << "couldn't find matching item for request: " << *request_obj
                   << "\n";
      return 1;
    }
    if (item->handled) {
      llvm::errs() << "request is already handled\n";
      return 1;
    }
    item->handled = true;
    auto response = createResponse(*item, *request_obj);

    // Write response data.
    llvm::SmallVector<char, 0> response_data;
    llvm::raw_svector_ostream(response_data) << response;
    auto response_size = response_data.size();
    auto response_header =
        llvm::support::endian::byte_swap<uint64_t, llvm::endianness::little>(
            response_size);
    ioSize = fwrite(&response_header, sizeof(response_header), 1, stdout);
    if (!ioSize) {
      llvm::errs() << "failed to write response header\n";
      return 1;
    }
    ioSize = fwrite(response_data.data(), response_size, 1, stdout);
    if (!ioSize) {
      llvm::errs() << "failed to write response data\n";
      return 1;
    }
    fflush(stdout);
  }
}

int _mock_plugin_main(const char *testSpectStr) {
  auto runner = TestRunner::create(testSpectStr);
  if (!runner) {
    return 1;
  }
  return runner->run();
}
